client/src/Utils/sigmfMetadata.ts

```diff
@@ -0,0 +1,271 @@
+    } else {
+      return `${origin.account}/${origin.container}/${origin.file_path}`;
+    }
+  }
+
+  getCanonicalFilePath() {
+    const origin = this.global['traceability:origin'];
+    if (origin) {
+      return `${origin.account}/${origin.container}/${origin.file_path}`;
+    }
+    return '';
+  }
+
+  getShortDescription() {
+    let shortDescription = this.global['core:description'] ?? '';
+    shortDescription = shortDescription.substring(0, 60);
+    if (shortDescription.length === 60) {
+      shortDescription += '...';
+    }
+    return shortDescription;
+  }
+
+  getThumbnailUrl() {
+    return this.getFullFilePath() + '.jpeg';
+  }
+  getDataUrl() {
+    return this.getFullFilePath() + '.sigmf-data';
+  }
+  getMetadataUrl() {
+    return this.getFullFilePath() + '.sigmf-meta';
+  }
+  getDataType() {
+    return this.global['core:datatype'];
+  }
+  getDataTypeDescription() {
+    return this.global['core:datatype']
+      .replace('c', 'complex\n')
+      .replace('r', 'real\n')
+      .replace('f', 'float\n')
+      .replace('i', 'signed int\n')
+      .replace('u', 'unsigned int\n')
+      .replace('8', '8 bits')
+      .replace('16', '16 bits')
+      .replace('32', '32 bits')
+      .replace('64', '64 bits')
+      .replace('_le', '');
+  }
+  getDescription() {
+    return this.global['core:description'] ?? '';
+  }
+
+  getLengthInMillionIQSamples() {
+    return (this.global['traceability:sample_length'] ?? 0) / 1e6;
+  }
+
+  getLengthInIQSamples() {
+    return this.global['traceability:sample_length'] ?? 0;
+  }
+
+  getLengthInBytes() {
+    return this.getLengthInIQSamples() * this.getBytesPerSample();
+  }
+
+  getCenterFrequency() {
+    for (let i = 0; i < this.captures.length; i++) {
+      if (this.captures[i]['core:frequency']) {
+        return this.captures[i]['core:frequency'];
+      }
+    }
+    return 0;
+  }
+}
+
+export class CaptureSegment {
+  'core:sample_start': number;
+  'core:global_index'?: number;
+  'core:header_bytes'?: number;
+  'core:frequency'?: number;
+  'core:datetime'?: string;
+  [key: string]: any;
+}
+
+export class TraceabilityOrigin {
+  type: string;
+  account: string;
+  container: string;
+  file_path: string;
+}
+
+export class Annotation {
+  'core:sample_start': number;
+  'core:sample_count'?: number;
+  'core:generator'?: string;
+  'core:label'?: string;
+  'core:comment'?: string;
+  'core:freq_lower_edge'?: number;
+  'core:freq_upper_edge'?: number;
+  'core:uuid'?: string;
+  'core:description'?: string;
+  capture_details?: string;
+
+  getAnnotationPosition(
+    lowerTile: number,
+    upperTile: number,
+    centerFrequency: number,
+    sampleRate: number,
+    fftSize: number,
+    zoomLevel: number
+  ) {
+    let result = {
+      x1: 0,
+      x2: 0,
+      y1: 0,
+      y2: 0,
+      visible: false,
+    };
+    let freq_lower_edge = this['core:freq_lower_edge'];
```

Comment:
how do we gracefully handle if these things aren't set?

Reply:
those fields are required for annotations. you cannot create an annotation object without them